forge-game/src/main/java/forge/game/GameAction.java

```diff
@@ -209,11 +209,27 @@ private Card changeZone(final Zone zoneFrom, Zone zoneTo, final Card c, Integer
             c.getController().setRevolt(true);
         }
 
+        Player putter = c.getController();
+
         // Don't copy Tokens, copy only cards leaving the battlefield
         // and returning to hand (to recreate their spell ability information)
         if (toBattlefield || (suppress && zoneTo.getZoneType().isHidden())) {
             copied = c;
 
+            // in some cases it's always affected that puts them in play (initially)
+            String defPutter;
+            if (cause == null || (!cause.hasParam("Putter") && (c.isToken() ||
```

Comment:
I don't know if I like it there

I probably would move the Putter logic into `AbilityKey.addCardZoneTableParams`

Reply:
hmn, the problem is the Putter can be different players during the same effect

and most effects currently don't need to modify moveParams between cards and reuse them instead

so I'd have to refactor those further, especially when they pass down to helpers in SpellAbilityEffect like with discard?

Reply:
imo it would be more clean if the stuff is done in the specific SpellAbilityEffect classes (or a helper function)

Reply:
@tool4ever can you move the Putter Logic into Effects instead?

Reply:
I tried but like I said above it got messy fast ;/
Not sure yet if I can find a cleaner approach